package/MDAnalysis/analysis/lineardensity.py

```diff
@@ -146,13 +150,19 @@ def _prepare(self):
         group = getattr(self._ags[0], self.grouping)
 
         # Get masses and charges for the selection
-        try:  # in case it's not an atom
-            self.masses = np.array([elem.total_mass() for elem in group])
-            self.charges = np.array([elem.total_charge() for elem in group])
-        except AttributeError:  # much much faster for atoms
+        if self.grouping == "atoms":
             self.masses = self._ags[0].masses
             self.charges = self._ags[0].charges
 
+        if self.grouping in ["residues", "segments"]:
+            self.masses = np.array([elem.atoms.total_mass() for elem in group])
+            self.charges = np.array(
+                [elem.atoms.total_charge() for elem in group])
+
+        if self.grouping == "fragments":
+            self.masses = np.array([elem.total_mass() for elem in group])
+            self.charges = np.array([elem.total_charge() for elem in group])
```

Comment:
So someone like @richardjgowers is probably going to tell me off for this, but the cost of accessing `.atoms.total_mass()` on an atomgroup is pretty much within error as calling `.total_mass()`.

Would it not make sense here to just use the same logic for residues, segments and fragments?